src/backend/cpu/sift.cpp

```diff
@@ -0,0 +1,1046 @@
+
+// maximum steps of keypoint interpolation before failure
+static const int MaxInterpSteps = 5;
+
+// default number of bins in histogram for orientation assignment
+static const int OriHistBins = 36;
+
+// determines gaussian sigma for orientation assignment
+static const float OriSigFctr = 1.5f;
+
+// determines the radius of the region used in orientation assignment */
+static const float OriRadius = 3.0f * OriSigFctr;
+
+// number of passes of orientation histogram smoothing
+static const int SmoothOriPasses = 2;
+
+// orientation magnitude relative to max that results in new feature
+static const float OriPeakRatio = 0.8f;
+
+// determines the size of a single descriptor orientation histogram
+static const float DescrSclFctr = 3.f;
+
+// threshold on magnitude of elements of descriptor vector
+static const float DescrMagThr = 0.2f;
+
+// factor used to convert floating-point descriptor to unsigned char
+static const float IntDescrFctr = 512.f;
+
+typedef struct
+{
+    float    f[4];
+    unsigned l[1];
+} feat_t;
+
+bool feat_cmp(feat_t i, feat_t j)
+{
+    for (int k = 0; k < 4; k++)
+        if (i.f[k] != j.f[k])
+            return (i.f[k] < j.f[k]);
+    if (i.l[0] != j.l[0])
+        return (i.l[0] < j.l[0]);
+
+    return true;
+}
+
+void array_to_feat(std::vector<feat_t>& feat, float *x, float *y, unsigned *layer, float *resp, float *size, unsigned nfeat)
+{
+    feat.resize(nfeat);
+    for (unsigned i = 0; i < feat.size(); i++) {
+        feat[i].f[0] = x[i];
+        feat[i].f[1] = y[i];
+        feat[i].f[2] = resp[i];
+        feat[i].f[3] = size[i];
+        feat[i].l[0] = layer[i];
+    }
+}
+
+template<typename T>
+void gaussian1D(T* out, const int dim, double sigma=0.0)
+{
+    if(!(sigma>0)) sigma = 0.25*dim;
+
+    T sum = (T)0;
+    for(int i=0;i<dim;i++)
+    {
+        int x = i-(dim-1)/2;
+        T el = 1. / sqrt(2 * PI_VAL * sigma*sigma) * exp(-((x*x)/(2*(sigma*sigma))));
+        out[i] = el;
+        sum   += el;
+    }
+
+    for(int k=0;k<dim;k++)
+        out[k] /= sum;
+}
+
+template<typename T>
+Array<T> gauss_filter(float sigma)
+{
+    // Using 6-sigma rule
+    unsigned gauss_len = (unsigned)round(sigma * 6 + 1) | 1;
+
+    Array<T> filter = createEmptyArray<T>(gauss_len);
+    gaussian1D((T*)getDevicePtr(filter), gauss_len, sigma);
+
+    return filter;
+}
+
+void gaussianElimination(float* A, float* b, float* x, const int n)
+{
+    // forward elimination
+    for (int i = 0; i < n-1; i++) {
+        for (int j = i+1; j < n; j++) {
+            float s = A[j*n+i] / A[i*n+i];
+
+            for (int k = i; k < n; k++)
+                A[j*n+k] -= s * A[i*n+k];
+
+            b[j] -= s * b[i];
+        }
+    }
+
+    for (int i = 0; i < n; i++)
+        x[i] = 0;
+
+    // backward substitution
+    float sum = 0;
+    for (int i = 0; i <= n-2; i++) {
+        sum = b[i];
+        for (int j = i+1; j < n; j++)
+            sum -= A[i*n+j] * x[j];
+        x[i] = sum / A[i*n+i];
+    }
+}
```

Comment:
For this implementation it will always be 3, but for other usages it could be any positive number I guess.